### PR TITLE
Replace inheritance_integrity by trait shenanigans 

### PR DESCRIPTION
**Pull request by nox**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15567)

### CODE REVIEW

File: components/domobject_derive/lib.rs [hunk 1]
```diff
@@ -14,17 +14,27 @@ pub fn expand_token_stream(input: proc_macro::TokenStream) -> proc_macro::TokenS
...
-        let first_field = fields.first().expect("#[derive(DomObject)] should not be applied on empty structs");
-        first_field.ident.as_ref().unwrap()
+    let fields = if let syn::Body::Struct(syn::VariantData::Struct(ref fields)) = type_.body {
+        fields
     } else {
         panic!("#[derive(DomObject)] should only be applied on proper structs")
     };
 
+    let first_field_name = fields
+        .first()
```
Comment (Ms2ger) [inline components/domobject_derive/lib.rs]: `split_first`?

Comment (nox) [inline components/domobject_derive/lib.rs]: For some reason I thought this one panicked, will amend.

Comment (nox) [inline components/domobject_derive/lib.rs]: Fixed.

**Comment by SimonSapin** [comment]
Is this covered by existing compile-fail tests?

---

Reviewed 1 of 1 files at r1, 3 of 4 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[components/domobject_derive/lib.rs, line 67 at r3](https://reviewable.io:443/reviews/servo/servo/15567#-Kd2LrLw3TDkftekhrAi:-Kd2LrLw3TDkftekhrAj:b-y05y6d) ([raw file](https://github.com/servo/servo/blob/1d8b2b530a4d4b61a2ec89570e2bddc301928dc0/components/domobject_derive/lib.rs#L67)):*
> ```Rust
>     items.append_all(field_types.iter().map(|ty| {
>         quote! {
>             impl #impl_generics ShouldNotImplDomObject for ((#params), #ty) #where_clause {}
> ```

This could use a code comment, in particular `((#params), #ty)`

---

*[components/plugins/lints/inheritance_integrity.rs, line 30 at r3](https://reviewable.io:443/reviews/servo/servo/15567#-Kd2I0uAc1dHiV5dsEuf:-Kd2I0uFWE2Xj4kGULXo:ba7sphp) ([raw file](https://github.com/servo/servo/blob/216a89f7766dd366c4afbeae42cf6e1fb4f67349/components/plugins/lints/inheritance_integrity.rs#L30)):*
> ```Rust
>         // Lints are run post expansion, so it's fine to use
>         // #[_dom_struct_marker] here without also checking for #[dom_struct]
>         if cx.tcx.has_attr(cx.tcx.hir.local_def_id(id), "_dom_struct_marker") {
> ```

Since this is removed, I think `_dom_struct_marker` can also be removed in `plugins/lib.rs` and `plugins/jstraceable.rs`.

---

*[components/plugins/lints/inheritance_integrity.rs, line 34 at r3](https://reviewable.io:443/reviews/servo/servo/15567#-Kd2ISwt3vYcegjRtCAp:-Kd2ISwuXl6om_nbHCqO:brn3ih0) ([raw file](https://github.com/servo/servo/blob/216a89f7766dd366c4afbeae42cf6e1fb4f67349/components/plugins/lints/inheritance_integrity.rs#L34)):*
> ```Rust
>             let reflector_span = def.fields().iter().enumerate()
>                                     .find(|&(ctr, f)| {
>                                         if match_lang_ty(cx, &*f.ty, "reflector") {
> ```

I think this was the last use of `match_lang_ty`, which can therefore be removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/15567#-:-Kd2N0DgE1UIbESp1MFC:b-n90w4h)*

**Comment by nox** [comment]
@SimonSapin Addressed your remarks.

**Comment by nox** [comment]
It's quite involved to make a test for it because you need to redefine the traits etc, given they are defined in `script` itself. It is also currently untested, so I am not sure it's worth making one as part of this PR.

**Comment by SimonSapin** [comment]
@bors-servo r+

---

Reviewed 5 of 5 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/15567#-:-Kd4FFV86M8jUtv56Fnh:bkg9nno)*

**Pull request merged**